File: 01/python/add_2020.py
#!/usr/bin/env python3

from typing import List, Tuple
import argparse
import textwrap
import sys


def add2020Two(num_list: List[int]) -> Tuple[int, int]:
    for i, num1 in enumerate(num_list[:-1]):
        for num2 in num_list[i+1:]:
            if num1 + num2 == 2020:
                return (num1, num2)
    raise ValueError("No valid results in num_list")


def add2020Three(num_list: List[int]) -> Tuple[int, int, int]:
    for i, num1 in enumerate(num_list[:-2]):
        for j, num2 in enumerate(num_list[i+1:-1]):
            for _, num3 in enumerate(num_list[i+j+1:]):
                if num1 + num2 + num3 == 2020:
                    return (num1, num2, num3)
    raise ValueError("No valid results in num_list")


def main() -> None:
    aparser = argparse.ArgumentParser(
        description=textwrap.dedent("""\
        Read a list of numbers from stdin and find the combination of two or
        three numbers add up to 2020 and multiply them together."""))
    aparser.add_argument("problem", choices=["one", "two"])
    args = aparser.parse_args()
    num_list = []
    for line in sys.stdin:
        num_list.append(int(line.strip()))
    if args.problem == "one":
        num1, num2 = add2020Two(num_list)
        print(num1, num2)
        print(num1 * num2)
    else:
        num1, num2, num3 = add2020Three(num_list)
        print(num1, num2, num3)
        print(num1 * num2 * num3)


if __name__ == "__main__":
    main()

File: 01/go/add_2020.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

var usage string = fmt.Sprintf("usage: %s [-h | --help] {one,two}\n", filepath.Base(os.Args[0]))

func add2020Two(numList []int) ([]int, error) {
	for i, num1 := range numList[0 : len(numList)-1] {
		for _, num2 := range numList[i+1:] {
			if num1+num2 == 2020 {
				return []int{num1, num2}, nil
			}
		}
	}
	return []int{-1, -1}, fmt.Errorf("Could not find any matching numbers")
}

func add2020Three(numList []int) ([]int, error) {
	for i, num1 := range numList[0 : len(numList)-2] {
		for j, num2 := range numList[i+1 : len(numList)-1] {
			for _, num3 := range numList[i+j+1:] {
				if num1+num2+num3 == 2020 {
					return []int{num1, num2, num3}, nil
				}
			}
		}
	}
	return []int{-1, -1, -1}, fmt.Errorf("Could not find any matching numbers")
}

func multiplySlice(numList []int) int {
	total := 1
	for _, n := range numList {
		total = total * n
	}
	return total
}

func readStdinOrDie() []int {
	numList := make([]int, 0)
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		num, err := strconv.Atoi(strings.TrimSpace(scanner.Text()))
		if err != nil {
			panic(err)
		}
		numList = append(numList, num)
	}
	if err_scan := scanner.Err(); err_scan != nil {
		panic(err_scan)
	}
	return numList
}

func printUsageAndDie() {
	fmt.Fprintf(os.Stderr, usage)
	os.Exit(1)
}

func printHelp() {
	fmt.Fprintf(os.Stdout, usage)
	fmt.Fprintln(os.Stdout, `Read a list of numbers from stdin and find the combination
	of two or three numbers add up to 2020 and multiply them together.`)
}

func main() {
	if len(os.Args) == 2 {
		if os.Args[1] == "-h" || os.Args[1] == "--help" {
			printHelp()
			os.Exit(0)
		}
		if os.Args[1] != "one" && os.Args[1] != "two" {
			printUsageAndDie()
		}
	} else {
		printUsageAndDie()
	}

	var func2020 func([]int) ([]int, error)
	switch os.Args[1] {
	case "one":
		func2020 = add2020Two
	case "two":
		func2020 = add2020Three
	}
	numList := readStdinOrDie()
	nums, err := func2020(numList)
	if err != nil {
		panic(err)
	}
	fmt.Println(nums)
	total := multiplySlice(nums)
	fmt.Println(total)
}
